Update methods used to fetch web extension data records to set an error if an issue occurred.

kiaraarose opened the pull request:
#### ff1c5f882081ba26030d9e273fc0dabec87aceff
<pre>
Update methods used to fetch web extension data records to set an error if an issue occurred.
<a href="https://bugs.webkit.org/show_bug.cgi?id=271998">https://bugs.webkit.org/show_bug.cgi?id=271998</a>
<a href="https://rdar.apple.com/124625910">rdar://124625910</a>

Reviewed by Timothy Hatcher.

We should notify apps if there was an issue in deleting or calculating extension storage.
If an error occurs, apps can decide to show an alert to their users. We add a test for this
unfortunately because I tested manually by forcing an error (e.g. result = makeUnexpected(&quot;error&quot;))
and I verified _WKWebExtensionDataRecord.errors was populated in Safari.

* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm:
(-[_WKWebExtensionSQLiteStore _deleteDatabase]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.h:
(NS_ERROR_ENUM):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.mm:
(WebKit::createDataRecordError):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecordInternal.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::getDataRecords):
(WebKit::WebExtensionController::getDataRecord):
(WebKit::WebExtensionController::removeData):
(WebKit::WebExtensionController::calculateStorageSize):
(WebKit::WebExtensionController::removeStorage):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDataRecordCocoa.mm:
Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecordInternal.h.
(WebKit::WebExtensionDataRecord::setError):
Sets an error on the corresponding _WKWebExtensionDataRecord.

* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
(-[_WKWebExtensionStorageSQLiteStore getValuesForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore getStorageSizeForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore getStorageSizeForAllKeysIncludingKeyedData:withCompletionHandler:]):
Add additional logging when an error occurs.

* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionDataRecord.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/277012@main">https://commits.webkit.org/277012@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/818ddb2140e8154aa82daa5db2f7518da20598ba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46441 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/25596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49041 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49116 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42482 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/29954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/23060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47019 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/29954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/19146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/29954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4486 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/29954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/41518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50950 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/21446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/23060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/22738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6474 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/22440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.h:
```diff
@@ -30,8 +30,29 @@
 
 #import <WebKit/_WKWebExtensionDataType.h>
 
+#define HAVE_UPDATED_WEB_EXTENSION_DATA_RECORD_ERROR_PROPERTIES 1
```
Is this needed? Can we use `respondsToSelector` for checking the availability?

xeenon replied:
Unfortunately no, when the methods change signature (like different completion handler args) then that isn't sufficient if the method was already in a previous SDK.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.h:
```diff
@@ -30,8 +30,29 @@
+ */
+typedef NS_ERROR_ENUM(_WKWebExtensionDataRecordErrorDomain, _WKWebExtensionDataRecordError) {
+    _WKWebExtensionDataRecordErrorUnknown,
+    _WKWebExtensionDataRecordErrorLocalStorageFailed,
+    _WKWebExtensionDataRecordErrorSessionStorageFailed,
+    _WKWebExtensionDataRecordErrorSyncStorageFailed,
```
I think we want to have either an `errors` array getter, that returns all applicable errors (my preference), or you only need one error code, since you have individual error properties, and `_WKWebExtensionDataRecordErrorLocalStorageFailed` will always be the case for `localStorageError`, etc.

Again, I think if we have a `errors` property, then these should stay as-is.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.h:
```diff
@@ -55,6 +76,15 @@ NS_SWIFT_NAME(_WKWebExtension.DataRecord)
+@property (nonatomic, nullable, readonly) NSError *sessionStorageError;
+
+/*! @abstract The error set if an issue occurred when either deleting or calculating sync storage. */
+@property (nonatomic, nullable, readonly) NSError *syncStorageError;
```
As mentioned, I think `@property (nonatomic, readonly, copy) NSArray<NSError *> *errors` will be better here since you have individual error codes per type already.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.mm:
```diff
@@ -144,6 +146,12 @@ - (unsigned long long)sizeOfDataTypes:(NSSet<_WKWebExtensionDataType> *)dataType
 
+NSError *createDataRecordError(_WKWebExtensionDataRecordError error, NSString *underlyingErrorMessage)
+{
+    NSDictionary *userInfo = underlyingErrorMessage ? @{ NSUnderlyingErrorKey: underlyingErrorMessage } : @{ };
```
`NSUnderlyingErrorKey` is for another `NSError`, not a `NSString`. You can use `NSDebugDescriptionErrorKey` for this. (You should rename the parameter to `debugDescription`.)

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -140,19 +140,31 @@ - (void)_didDeleteLocalStorage:(NSNotification *)notification
+
+                Ref record = recordHolder->recordsMap.ensure(uniqueIdentifier, [&] {
+                    return WebExtensionDataRecord::create(displayName, uniqueIdentifier);
+                }).iterator->value;
```
This should likely move out of the block, so you don't duplicate it from inside `calculateStorageSize`.

kiaraarose replied:
there's a `continue` below so we wouldn't hit the `calculateStorageSize` block.

xeenon replied:
Right, but it can still be shared code with the code on line 161-163. You should only need to do this in one place.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -188,14 +200,25 @@ - (void)_didDeleteLocalStorage:(NSNotification *)notification
+
+            Ref record = recordHolder->recordsMap.ensure(matchingUniqueIdentifier, [&] {
+                return WebExtensionDataRecord::create(displayName, matchingUniqueIdentifier);
+            }).iterator->value;
```
Ditto on moving outside the block to share with `calculateStorageSize`.

xeenon replied:
This code is still duplicated in two places in this method.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecordInternal.h:
```diff
@@ -44,12 +44,17 @@ template<> struct WrapperTraits<WebExtensionDataRecord> {
 
+@property (readwrite, setter=_setLocalStorageError:) NSError *localStorageError;
+@property (readwrite, setter=_setSessionStorageError:) NSError *sessionStorageError;
+@property (readwrite, setter=_setSyncStorageError:) NSError *syncStorageError;
```
These should not need private setters. (See `WebExtensionDataRecord::setError`.)

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDataRecordCocoa.mm:
```diff
@@ -0,0 +1,58 @@
+        break;
+    case WebExtensionDataType::Sync:
+        wrapper().syncStorageError = createDataRecordError(_WKWebExtensionDataRecordErrorSyncStorageFailed, underlyingErrorMessage);
+        break;
```
This seems backwards. The ObjC object should get the error from the C++ object. The C++ object should not be setting the object on the ObjC wrapper.

In this case `WebExtensionDataRecord` should have an `NSArray` of errors, similar to `WebExtension::errors()`.

kiaraarose replied:
spoke about this offline with @xeenon. it's best to have WebExtensionDataRecord mirror _WKWebExtensionDataRecord. ideally, for all the web extension classes we've written, the c++ object should be a source of truth. with this approach, the _WKWebExtensionDataRecord would have these errors but the WebExtensionDataRecord wouldn't, which is not the behavior we want to promote.

szewai commented on Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
```diff
@@ -108,10 +108,9 @@ class WebExtensionController : public API::ObjectImpl<API::Object::Type::WebExte
 
-    void calculateStorageSize(_WKWebExtensionStorageSQLiteStore *, WebExtensionDataType, CompletionHandler<void(size_t)>&&);
-    void removeStorage(_WKWebExtensionStorageSQLiteStore *, WebExtensionDataType, CompletionHandler<void()>&&);
+    void calculateStorageSize(_WKWebExtensionStorageSQLiteStore *, WebExtensionDataType, CompletionHandler<void(size_t, NSString *errorMessage)>&&);
```
This should probably be `CompletionHandler<void(Expected<size_t, WebExtensionError>&&)>&&` if size_t and error message cannot appear at the same time.

kiaraarose replied:
`getStorageSizeForKeys` will actually return 0 if there was an error calculating the storage size, which would be passed in here

xeenon replied:
But 0 can be assumed with any error, so `Expected<size_t, WebExtensionError>` would be best here.

kiaraarose replied:
will do! thanks @szewai for the suggestion

szewai commented on Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
```diff
@@ -108,10 +108,9 @@ class WebExtensionController : public API::ObjectImpl<API::Object::Type::WebExte
-    void calculateStorageSize(_WKWebExtensionStorageSQLiteStore *, WebExtensionDataType, CompletionHandler<void(size_t)>&&);
-    void removeStorage(_WKWebExtensionStorageSQLiteStore *, WebExtensionDataType, CompletionHandler<void()>&&);
+    void calculateStorageSize(_WKWebExtensionStorageSQLiteStore *, WebExtensionDataType, CompletionHandler<void(size_t, NSString *errorMessage)>&&);
+    void removeStorage(_WKWebExtensionStorageSQLiteStore *, WebExtensionDataType, CompletionHandler<void(const String& errorMessage)>&&);
```
This should probably be `CompletionHandler<void(Expected<void, WebExtensionError>&&)>&&`

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -211,34 +233,47 @@ - (void)_didDeleteLocalStorage:(NSNotification *)notification
+        if (errorMessage)
+            completionHandler(makeUnexpected(errorMessage));
+        else
+            completionHandler({ });
```
should be passing in the size here. will update.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDataRecordCocoa.mm:
```diff
@@ -0,0 +1,61 @@
+
+namespace WebKit {
+
+void WebExtensionDataRecord::setError(NSString *debugDescription, WebExtensionDataType type)
```
`addError` would be a better name.

kiaraarose committed:
Update methods used to fetch web extension data records to set an error if an issue occurred.
https://bugs.webkit.org/show_bug.cgi?id=271998
rdar://124625910

Reviewed by Timothy Hatcher.

We should notify apps if there was an issue in deleting or calculating extension storage.
If an error occurs, apps can decide to show an alert to their users. We add a test for this
unfortunately because I tested manually by forcing an error (e.g. result = makeUnexpected("error"))
and I verified _WKWebExtensionDataRecord.errors was populated in Safari.

* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm:
(-[_WKWebExtensionSQLiteStore _deleteDatabase]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.h:
(NS_ERROR_ENUM):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.mm:
(WebKit::createDataRecordError):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecordInternal.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::getDataRecords):
(WebKit::WebExtensionController::getDataRecord):
(WebKit::WebExtensionController::removeData):
(WebKit::WebExtensionController::calculateStorageSize):
(WebKit::WebExtensionController::removeStorage):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDataRecordCocoa.mm:
Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecordInternal.h.
(WebKit::WebExtensionDataRecord::setError):
Sets an error on the corresponding _WKWebExtensionDataRecord.

* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
(-[_WKWebExtensionStorageSQLiteStore getValuesForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore getStorageSizeForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore getStorageSizeForAllKeysIncludingKeyedData:withCompletionHandler:]):
Add additional logging when an error occurs.

* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionDataRecord.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/277012@main
